synapse/lib/lmdbslab.py

```diff
@@ -1541,109 +1512,124 @@ def bump(self):
             self.curs.close()
             self.bumped = True
 
+    def item(self):
+        return self.curs.item()
+
+    def iterfunc(self, curs):
+        return curs.iternext()
+
+    def resume(self, item):
+
+        if not self.dupsort:
+            return self.curs.set_range(item[0])
+
+        if self.curs.set_range_dup(*item):
+            return True
+
+        if not self.curs.set_range(item[0]):
+            return False
+
+        # if the key is the same, we're at a previous
+        # entry and need to skip dups to the next key
+        if self.curs.key() == item[0]:
+            return self.curs.next_nodup()
+
+        return True
+
+class ScanKeys(Scan):
+
+    def item(self):
+        return self.curs.key()
+
+    def iterfunc(self, curs):
+        return self.curs.iternext(keys=True, values=False)
+
+    def resume(self, item):
+        return self.curs.set_range(item)
+
 class ScanBack(Scan):
     '''
     A state-object used by Slab.  Not to be instantiated directly.
 
     Scans backwards.
     '''
-    def __init__(self, slab, db, singlekey=False):
-        Scan.__init__(self, slab, db, singlekey)
-        if self.singlekey:
-            self.iterfunc = functools.partial(lmdb.Cursor.iterprev_dup, keys=True)
-        else:
-            self.iterfunc = lmdb.Cursor.iterprev
+    def iterfunc(self, curs):
+        return curs.iterprev()
 
     def first(self):
 
         if not self.curs.last():
             return False
 
-        self.genr = self.curs.iterprev()
+        if self.dupsort:
+            self.curs.last_dup()
```

Comment:
I believe `last` should already to this if the db is opened with dupsort=True

Reply:
Thanks, i'll test.